### DISCUSSION TITLE
[Bug] LazyView always loads the BindingContext

### DISCUSSION THREAD

**Discussion by auriou** [xamarin/XamarinCommunityToolkit#1408] [Q&A] [answered]
### Description
Following this example into XCT.Sample
https://github.com/xamarin/XamarinCommunityToolkit/blob/develop/samples/XCT.Sample/Pages/Views/TabView/LazyTabPage.xaml

Each time we add a BindingContext to a LazyView, the viewmodel is created at the start,  without Lazy mode
On the other hand the xaml view is well constructed just when it is visible.

### Steps to Reproduce

1. Into XCT.Sample
2. put a breakpoint on the constructor of LazyTestViewModel class
3. Goto View / TabView / LazyLoading Tab
4. The breakpoint will be reached upon loading without the second tab being clicked
5. All the ViewModels of the lazyView will be created on the first load, which is not normal in my opinion

- Version with issue: actual main branch

**Comment by pictos** [comment]
@auriou in that case the VM is created because it's a static value, as you can [see here](https://github.com/xamarin/XamarinCommunityToolkit/blob/70d5df595713e8556b0d52c8c3db724439235569/samples/XCT.Sample/ViewModels/Views/Tabs/NormalTestViewModel.cs#L7) and it's the used in the [NormalTabs](https://github.com/xamarin/XamarinCommunityToolkit/blob/70d5df595713e8556b0d52c8c3db724439235569/samples/XCT.Sample/Pages/Views/TabView/NormalTestView.xaml.cs#L11).

So for this sample, the VM will be created even if the LazyView isn't loaded, because it's the same view model used by Non-LazyView. And also I'm referencing a static property, so the VM will be instantiated ASAP.

Hope that makes sense for you

**Comment by auriou** [comment]
Ok, I see, so we must avoid using the BindingContext property in the Xaml for LazyView, as we can bind BindingContext only on static code

**Reply by pictos** [reply]
I need to check, but I would say that you can do
```xml
<xct:LazyView x:TypeArguments="local:LazyTestView" IsLoaded="{Binding Loaded}" >
         <xct:LazyView.BindingContext>
                  <VM:YourViewModel/>
         </xct:LazyView.BindingContext>
</xct:LazyView >
```

and (maybe) have the desired behavior

**Reply by auriou** [reply]
@pictos It's a shame but it's the same

BindingContext of the BindableObject class in Xamarin.Form would need to implement a method

```c#
protected override void OnBindingContextChanging()
```
there is an overload when the object has changed but unfortunately not before it changes
```c#
protected override void OnBindingContextChanged()
```

the best solution if you want to make the viewModel lazy, you have to declare it in the xml or cs code of the view that the LazyView will load

**Comment by auriou** [comment, marked as answer]
I solved my problem by implementing a class that inherits from LazyView

```c#
public class CustomLazyView<TView> : LazyView<TView> where TView : View, new()
{
    public static readonly BindableProperty ViewModelProperty =
        BindableProperty.Create(nameof(ViewModel), typeof(Type),
            typeof(CustomLazyView<TView>), default);

    public Type ViewModel
    {
        get => (Type)GetValue(ViewModelProperty);
        set => SetValue(ViewModelProperty, value);
    }

    public override ValueTask LoadViewAsync()
    {
        if (ViewModel != null)
            BindingContext = Resolver.Resolve(ViewModel);

        return base.LoadViewAsync();
    }
}
```

```xaml
<xct:TabViewItem Style="{StaticResource TabViewItemShell}" Text="{x:Static styles:IconsMaterial.Account}">
    <controls:CustomLazyView x:TypeArguments="views:SettingPage" ViewModel="{x:Type viewModels:SettingViewModel}" />
</xct:TabViewItem>

```
